### PR TITLE
JIRA: VZ-5488 support upgrades with ephemeral storage or authproxy transition 

### PR DESCRIPTION
**Pull request by jmaron99**
# Description

This PR addresses two upgrade scenarios:
- An upgrade with ephemeral storage from VZ 1.0.x to 1.3+
- An upgrade from authproxy as resource in VZ chart to a version in which authproxy is installed by a separate chart

Specific Modifications:
- VZ component disassociates authproxy in upgrade code prior to proceeding with upgrading.  It also marks the authproxy associated resources as resources to keep (rather than delete)
- The Authproxy component removes the "keep" resource policy annotation since it wants to manage those resources in all instances (install and upgrade)
- VZ component modified to only run the ES post upgrade test if non-ephemeral storage in use (ephemeral storage during upgrade prevents the retention of the index between upgrades)

Fixes VZ-5488

# Checklist 

As the author of this PR, I have:

- [x] Checked that I included or updated copyright and license notices in all files that I altered
- [x] Added or updated unit tests for any new functions I added
- [ ] Added or updated integration tests if appropriate
- [x] Added or updated acceptance tests if appropriate

Code reviewer, please confirm this PR:

- [ ] Addressed the requirement and meets the acceptance criteria
- [ ] Does not introduce unrelated or spurious changes
- [ ] Does not introduce any unapproved dependency
- [ ] Makes sense and it easy to understand, and/or difficult areas of code are clearly documented so that they can be understood

### CODE REVIEW

File: platform-operator/controllers/verrazzano/component/authproxy/authproxy.go [hunk 1]
```diff
@@ -118,6 +126,77 @@ func AppendOverrides(ctx spi.ComponentContext, _ string, _ string, _ string, kvs
...
+func authproxyPreHelmOps(ctx spi.ComponentContext) error {
+	return reassociateResources(ctx.Client())
+}
+
+//reassociateResources updates the resources to ensure they are managed by this release/component.  The resource policy
+// annotation is removed to ensure that helm manages the lifecycle of the resources (the resource policy annotation is
+// added to ensure the resources are disassociated from the VZ chart which used to manage these resources)
+func reassociateResources(cli clipkg.Client) error {
+	authproxyReleaseName := types.NamespacedName{Name: ComponentName, Namespace: ComponentNamespace}
+	namespacedName := authproxyReleaseName
```
Comment (anders-swanson) [inline platform-operator/controllers/verrazzano/component/authproxy/authproxy.go:139]: what's the point of assigning authPorxyReleaseName to namespacedName?

File: platform-operator/controllers/verrazzano/component/authproxy/authproxy.go [hunk 2]
```diff
@@ -118,6 +126,77 @@ func AppendOverrides(ctx spi.ComponentContext, _ string, _ string, _ string, kvs
 	return kvs, nil
 }
 
+// authproxyPreHelmOps performs any operations required prior to the upgrade
```
Comment (anders-swanson) [inline platform-operator/controllers/verrazzano/component/authproxy/authproxy.go:129]: comment should describe what operations are done?

File: platform-operator/controllers/verrazzano/component/verrazzano/verrazzano.go [hunk 3]
```diff
@@ -510,16 +515,20 @@ func fixupElasticSearchReplicaCount(ctx spi.ComponentContext, namespace string)
...
 	ctx.Log().Debugf("Elasticsearch Post Upgrade: Output of the health of the Elasticsearch cluster %s", string(output))
-	// If the data node count is seen as 1 then the node is considered as single node cluster
-	if strings.Contains(string(output), `"number_of_data_nodes":1,`) {
-		// Login to Elasticsearch and update index settings for single data node elasticsearch cluster
-		putCmd := execCommand("kubectl", "exec", pod.Name, "-n", namespace, "-c", containerName, "--", "sh", "-c",
-			fmt.Sprintf(`curl -v -XPUT -d '{"index":{"auto_expand_replicas":"0-1"}}' --header 'Content-Type: application/json' -s -k --fail http://localhost:%d/%s/_settings`, httpPort, indexPattern))
-		_, err = putCmd.Output()
-		if err != nil {
-			return ctx.Log().ErrorfNewErr("Failed in Elasticsearch post-upgrade: Error logging into Elasticsearch: %v", err)
+	if ctx.EffectiveCR().Spec.DefaultVolumeSource != nil && ctx.EffectiveCR().Spec.DefaultVolumeSource != nil {
```
Comment (anders-swanson) [inline platform-operator/controllers/verrazzano/component/verrazzano/verrazzano.go:518]: checks same var for nil twice

File: platform-operator/controllers/verrazzano/component/authproxy/authproxy.go [hunk 4]
```diff
@@ -118,6 +126,77 @@ func AppendOverrides(ctx spi.ComponentContext, _ string, _ string, _ string, kvs
...
+	}
+
+	// namespaced resources
+	for _, obj := range objects {
+		if _, err := removeResourcePolicyAnnotation(cli, obj, namespacedName); err != nil {
+			return err
+		}
+	}
+
+	// additional namespaced resources managed by this helm chart
```
Comment (anders-swanson) [inline platform-operator/controllers/verrazzano/component/authproxy/authproxy.go:159]: if you want you could do all of these in the loop by creating a wrapping struct that is 
struct {
 obj
 namespacedName
}

and creating a list of those wrappers to run through in the loop.

File: platform-operator/controllers/verrazzano/component/verrazzano/verrazzano.go (hunk 3 above)
Comment (jmaron99) [reply platform-operator/controllers/verrazzano/component/verrazzano/verrazzano.go:518]: Second one is supposed to be EmptyDir.  Not sure how that got lost.

File: platform-operator/controllers/verrazzano/component/authproxy/authproxy.go (hunk 4 above)
Comment (jmaron99) [reply platform-operator/controllers/verrazzano/component/authproxy/authproxy.go:159]: That does sound better but I have just today to get this merged and I'm not sure if I have the cycles to run through all the testing again to ensure no slip up.

File: platform-operator/controllers/verrazzano/component/authproxy/authproxy.go (hunk 1 above)
Comment (jmaron99) [reply platform-operator/controllers/verrazzano/component/authproxy/authproxy.go:139]: remnant from a previous iteration. will remove.

File: platform-operator/controllers/verrazzano/component/verrazzano/verrazzano.go (hunk 3 above)
Comment (jmaron99) [reply platform-operator/controllers/verrazzano/component/verrazzano/verrazzano.go]: This one is going to require me to run multiple ephemeral storage upgrade tests to ensure it's working correctly.  Will update here when those are completed.

File: platform-operator/controllers/verrazzano/component/authproxy/authproxy.go (hunk 4 above)
Comment (jmaron99) [reply platform-operator/controllers/verrazzano/component/authproxy/authproxy.go:172]: Actually went ahead and implemented that

**Review by anders-swanson — APPROVED** [review]
LGTM

File: platform-operator/controllers/verrazzano/component/authproxy/authproxy.go (hunk 4 above)
Comment (anders-swanson) [reply platform-operator/controllers/verrazzano/component/authproxy/authproxy.go:172]: I was thinking something along the lines of
```
type namespacedObject struct {
  name types.NamespacedName
  obj     controllerutil.object
}

objs = []namespcedObject{
  &corev1.Service{}, types.NamespacedName{...},
  ...
}

for _, obj := range objs {
  reassociate(obj)
}
```

**Pull request merged**